[Fix-Proxy] Fix in memory caching memory leak

ishaan-jaff opened the pull request:
## Title

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Feature
🐛 Bug Fix
🧹 Refactoring
📖 Documentation
🚄 Infrastructure
✅ Test

## Changes

<!-- List of changes -->

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locall
If UI changes, send a screenshot/GIF of working UI fixes

<!-- Test procedure -->

msabramo committed:
Turn on message logging via request header

msabramo committed:
Rename request header

msabramo committed:
Document feature

ishaan-jaff committed:
feat -  add debug_utils

ishaan-jaff committed:
fix - clean up in memory cache

Commit:
fix(router.py): check if azure returns 'content_filter' response + fallback available -> fallback

Exception maps azure content filter response exceptions

ishaan-jaff committed:
fix caching clear in memory cache mem util

ishaan-jaff committed:
Merge branch 'main' into litellm_fix_in_mem_usage

ishaan-jaff committed:
fix in mem cache tests

ishaan-jaff commented:
Great suggestion @Manouchehri i think we can do it in this PR itself

Commit:
fix(vertex_httpx.py): ignore vertex finish reason - wait for stream to end

Fixes https://github.com/BerriAI/litellm/issues/4339

Commit:
fix(vertex_httpx.py): flush remaining chunks from stream

Commit:
fix(add-exception-mapping-+-langfuse-exception-logging-for-streaming-exceptions): add exception mapping + langfuse exception logging for streaming exceptions

Fixes https://github.com/BerriAI/litellm/issues/4338

Commit:
bump: version 1.40.24 → 1.40.25

Commit:
docs(team_budgets.md): cleanup docs

Commit:
fix(redact_messages.py): fix get

Commit:
fix(test_dynamic_rate_limit_handler.py): cleanup

Commit:
fix(redact_messages.py): fix pr

Commit:
test(test_proxy_server_langfuse.py): cleanup tests causing OOM issues.

Commit:
test(test_completion.py): handle replicate api error

Commit:
test: skip unstable tests

jeromeroussin commented on litellm/caching.py:
```diff
@@ -139,8 +152,23 @@ async def async_increment(self, key, value: float, **kwargs) -> float:
+        - loop through all keys in cache, check if they are expired
+        - if yes, delete them
+        """
+        self.cache_dict = {}
```
I am surprised you don't try to keep some entries around. Maybe I misunderstand the code, I imagined you'd need to keep entries for the current minute at least.

ishaan-jaff replied:
fixed to use a custom eviction policy 

```python
def evict_cache(self):
        """
        Eviction policy:
        - check if any items in ttl_dict are expired -> remove them from ttl_dict and cache_dict
        This guarantees the following:
        - 1. When item ttl not set: At minimumm each item will remain in memory for 5 minutes
        - 2. When ttl is set: the item will remain in memory for at least that amount of time
        - 3. the size of in-memory cache is bounded
        """
        for key in list(self.ttl_dict.keys()):
            if time.time() > self.ttl_dict[key]:
                self.cache_dict.pop(key, None)
                self.ttl_dict.pop(key, None)
```

jeromeroussin commented on litellm/caching.py:
```diff
@@ -64,10 +64,16 @@ async def disconnect(self):
 
 
 class InMemoryCache(BaseCache):
```
Does the cache in Redis keep growing and need regular cleanup?

ishaan-jaff replied:
yes it does - perhaps we can set a default ttl on redis caching. Planning on addressing that in a separate PR

corrm committed:
chore: Improved prompt generation in ollama_pt function

corrm committed:
chore: Improved OllamaConfig get_required_params and ollama_acompletion and ollama_async_streaming functions

corrm committed:
Added improved function name handling in ollama_async_streaming

corrm committed:
Rename ollama prompt 'Function' word to 'Name'

7HR4IZ3 committed:
fix: Lunary integration

Fixes the bug of litellm not logging system messages to lunary

ishaan-jaff committed:
fix ui login bug

ishaan-jaff committed:
docs - update telemetry

Commit:
fix(utils.py): fix exception_mapping check for errors

If exception already mapped - don't attach traceback to it

ishaan-jaff committed:
fix use caching lib

ishaan-jaff committed:
fix InMemoryCache

ishaan-jaff committed:
fix router.py

ishaan-jaff committed:
fix testing env

ishaan-jaff committed:
use lru cache

ishaan-jaff committed:
cleanup InMemoryCache

Commit:
fix(router.py): use user-defined model_input_tokens for pre-call filter checks

ishaan-jaff committed:
fix install on python 3.8

Commit:
test(test_router.py): add testing

ishaan-jaff committed:
fix python3.8 with cachetools

ishaan-jaff committed:
use cache tools as dep

ishaan-jaff committed:
feat - allow user to define public routes

ishaan-jaff committed:
example config with public routes

Commit:
fix(router.py): log rejected router requests to langfuse

Fixes issue where rejected requests weren't being logged

ishaan-jaff committed:
feat - refactor /spend/calculate

Commit:
fix(vertex_httpx.py): cover gemini content violation (on prompt)

Commit:
fix(vertex_httpx.py): Return empty model response for content filter violations

ishaan-jaff committed:
test - spend/calculate endpoints

Commit:
fix(utils.py): catch 422-status errors

ishaan-jaff committed:
add helper to check route_in_additonal_public_routes

ishaan-jaff committed:
example cofnig with public routes

ishaan-jaff committed:
fix importing litellm

ishaan-jaff committed:
feat use custom eviction policy

ishaan-jaff committed:
fix config.yaml

ishaan-jaff committed:
fix deps

ishaan-jaff committed:
docs control available public routes

ishaan-jaff committed:
test - aliases on /spend/calculate

ishaan-jaff committed:
/spend/calculate use model aliases on this endpoint

ishaan-jaff committed:
bump: version 1.40.25 → 1.40.26

ishaan-jaff committed:
fix default ttl for InMemoryCache

Commit:
fix(langfuse.py): cleanup

Commit:
docs(routing.md): add quickstart

corrm committed:
Rename ollama prompt:
- 'Function' word to 'FunctionName'
- 'Tool Call' to `FunctionCall`
- 'Tool Call Result' to 'FunctionCall Result'

_I found that changes make some models better_

corrm committed:
Improve ollama prompt: this formula give good result with AutoGen

paul-gauthier committed:
Added openrouter/anthropic/claude-3.5-sonnet to model json

paul-gauthier committed:
Added openrouter/anthropic/claude-3-haiku-20240307

ishaan-jaff committed:
ci/cd add debugging for cache eviction

kiriloman committed:
Fix Groq prices

ishaan-jaff committed:
add nvidia nim to __init__

ishaan-jaff committed:
feat - add nvidia nim to main.py

lolsborn committed:
create litellm user to fix issue in k8s where prisma fails due to user nobody without home directory

ishaan-jaff committed:
feat - add param mapping for nvidia nim

ishaan-jaff committed:
test - nvidia nim

Commit:
Update README.md

Commit:
Update README.md

ishaan-jaff committed:
docs - add nvidia nim

ishaan-jaff committed:
docs nvidia_nim

ishaan-jaff committed:
ci/cd run again

ishaan-jaff committed:
feat - use n in mock completion

ishaan-jaff committed:
test - test_mock_request_n_greater_than_1

Commit:
feat(router.py): support mock testing content policy + context window fallbacks

Commit:
docs(reliability.md): add doc on mock testing fallbacks

ishaan-jaff committed:
fix using mock completion

Commit:
fix(router.py): improve error message returned for fallbacks

Commit:
fix(utils.py): add coverage for anthropic content policy error - vertex ai

Commit:
fix(utils.py): predibase exception mapping - map 424 as a badrequest error

ishaan-jaff committed:
fix - verify license without api request

Commit:
fix(predibase.py): support json schema on predibase

ishaan-jaff committed:
feat - add secret detection

ishaan-jaff committed:
init secret detection callback

ishaan-jaff committed:
fix only use crypto imports when needed

ishaan-jaff committed:
fix async_pre_call_hook

Commit:
fix(router.py): set `cooldown_time:` per model

Commit:
docs(routing.md): add dynamic cooldowns to docs

ishaan-jaff committed:
clean up secret detection

ishaan-jaff committed:
test secret detection

ishaan-jaff committed:
docs - secret detection

ishaan-jaff committed:
fix detect secrets test

ishaan-jaff committed:
bump: version 1.40.26 → 1.40.27

Commit:
Revert "Create litellm user to fix issue with prisma in k8s "

ishaan-jaff committed:
run again

Commit:
docs(function_call.md): cleanup

ishaan-jaff committed:
feat - add fireworks ai config for param mapping

ishaan-jaff committed:
fix fireworks ai config

ishaan-jaff committed:
add fireworks ai param mapping

ishaan-jaff committed:
test fireworks ai tool calling

ishaan-jaff committed:
fix +  test fireworks ai param mapping for tools

ishaan-jaff committed:
test - fireworks ai param mapping

Commit:
docs(reliable_completions.md): improve headers for easier searching

ishaan-jaff committed:
add gemini-1.0-ultra-001

ishaan-jaff committed:
fix gemini ultra info

guitard0g committed:
Add return type annotations to util types

ishaan-jaff committed:
add vertex text-bison

ishaan-jaff committed:
add chat-bison-32k@002

ishaan-jaff committed:
add code-bison

ishaan-jaff committed:
add code-gecko-latest

ishaan-jaff committed:
add codechat-bison@latest

ishaan-jaff committed:
forward otel traceparent in request headers

ishaan-jaff committed:
fix add ollama codegemma

ishaan-jaff committed:
fix cost tracking for whisper

ishaan-jaff committed:
test_spend_logs_payload_whisper

ishaan-jaff committed:
Revert "Add return type annotations to util types"

This reverts commit faef56fe696ff3eba0fcff80c3270534b2887648.

ishaan-jaff committed:
fix handle_openai_chat_completion_chunk

ishaan-jaff committed:
fix - reuse client initialized on proxy config

Commit:
fix(bedrock_httpx.py): Fix https://github.com/BerriAI/litellm/issues/4415

ishaan-jaff committed:
add volcengine as provider to litellm

ishaan-jaff committed:
add initial support for volcengine

ishaan-jaff committed:
docs - volcengine

ishaan-jaff committed:
test volcengine

ishaan-jaff committed:
add codestral pricing

ishaan-jaff committed:
add source for codestral pricing

ishaan-jaff committed:
vertex testing

ishaan-jaff committed:
fix gemini test

ishaan-jaff committed:
bump: version 1.40.27 → 1.40.28

ishaan-jaff committed:
ci/cd run again

Commit:
docs(openai_compatible.md): doc on disabling system messages

Commit:
fix(utils.py): add new special token for cleanup

djliden committed:
Update databricks.md

updates some references to predibase to refer to Databricks

djliden committed:
Update databricks.md

fixes a couple of examples to use correct endpoints/point to correct models

Commit:
fix(utils.py): handle arguments being None

Fixes https://github.com/BerriAI/litellm/issues/4440

Commit:
bump: version 1.40.28 → 1.40.29

djliden committed:
undoes changes to proxy yaml api key/base

ishaan-jaff committed:
docs - fix model name on claude-3-5-sonnet-20240620 anthropic

ishaan-jaff committed:
fix raise better error message on reaching failed vertex import

ishaan-jaff committed:
fix secret redaction logic

ishaan-jaff committed:
test - test_chat_completion_request_with_redaction

ishaan-jaff committed:
feat - improve secret detection

ishaan-jaff committed:
add stricter secret detection

ishaan-jaff committed:
fix secret scanner

ishaan-jaff committed:
fix error message on v2/model info

ishaan-jaff committed:
test fix secret detection

ishaan-jaff committed:
ci/cd run again

ishaan-jaff committed:
remove debug print statement

ishaan-jaff committed:
Merge branch 'main' into litellm_fix_in_mem_usage
